src/v/v8_engine/script.cc

```diff
@@ -30,16 +39,46 @@ script::script(size_t max_heap_size_in_bytes) {
       0, max_heap_size_in_bytes);
     _isolate = std::unique_ptr<v8::Isolate, isolate_deleter>(
       v8::Isolate::New(isolate_params), isolate_deleter());
+
+    _watchdog.set_callback([this] { stop_execution(); });
 }
 
 script::~script() {
     _function.Reset();
     _context.Reset();
 }
 
-void script::init(std::string_view name, ss::temporary_buffer<char> js_code) {
-    compile_script(std::move(js_code));
-    set_function(name);
+ss::future<>
+script::init(ss::sstring name, ss::temporary_buffer<char> js_code) {
+    return ss::with_semaphore(
+      _mutex,
+      1,
+      [this, js_code = std::move(js_code), name = std::move(name)]() mutable {
+          return _executor
+            .submit(
+              [this] {
+                  _is_cancel = false;
```

Comment:
why must `is_cancel` always be set to false? I feel like if the system has shutdown, and you try to execute , you can throw.

Reply:
is_cancel means that watchdog stopped v8::script due to timeout. It is not about system, it only about one execution. Maybe better use ss::so_with(false) in run function

Reply:
maybe we can do better, i don't like globals , i like your idea about a local bool within the context of the function , its nicer yes

Reply:
Oh, I can not use local is_cancel flag. I need to pass it to timer callback.
If I want use local function I need to use local timer. I do not know is it good idea to construct timer for each run call?

Reply:
What about this, we give the responsibility of all of this timer stuff to the executor. The executor has one timer, and when it fires, we check the queue for items which could be expired. We tag each task with a timestamp, and cancel tasks that have timestamps that have exceeded a threshold. So the `script` class just configures the `execturor` with a max_timeout (it can pass this value to the constructor or it can be a global const), and calls to `submit` just throw the timeout exception if it occurs. No booleans and no timers in script.h

Reply:
Then in the exception handler for `timeout_exception` you can call `_isolate->TerminateExecution()`

Reply:
looks like you want use std::thread for timeouts. I do not know is it good to use 1 thread in executor only for timeouts

Reply:
Done